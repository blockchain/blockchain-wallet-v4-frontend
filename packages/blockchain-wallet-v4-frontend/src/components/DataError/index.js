import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { FormattedMessage } from 'react-intl'
import {
  Image,
  Link,
  Text,
  TextGroup,
  Button
} from 'blockchain-info-components'
import { checkForVulnerableAddressError } from 'services/ErrorCheckService'

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
`
const Empty = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin-top: 20px;
`
const Header = styled(Text)`
  margin-top: 30px;
  margin-bottom: 20px;
`
const MessageText = styled(Text)`
  width: 80%;
  margin-bottom: 20px;
`

const DataErrorContainer = props => {
  const { message, onClick, onArchive } = props
  const e2e = props['data-e2e']
  const renderErrorHandling = msg => {
    const vulnerableAddress = checkForVulnerableAddressError(msg)
    if (vulnerableAddress) {
      return (
        <React.Fragment>
          <MessageText size='18px' weight={300}>
            {msg}
          </MessageText>
          <Button nature='primary' onClick={() => onArchive(vulnerableAddress)}>
            <Text size='18px' weight={300} color='white'>
              <FormattedMessage
                id='components.dataerror.archiveaddress'
                defaultMessage='Archive Address'
              />
            </Text>
          </Button>
        </React.Fragment>
      )
    } else {
      return (
        <TextGroup inline>
          <Text size='18px' weight={300}>
            <FormattedMessage
              id='components.dataerror.body'
              defaultMessage='Please '
            />
          </Text>
          <Link
            size='18px'
            data-e2e={e2e ? `${e2e}Link` : ''}
            onClick={onClick}
          >
            <FormattedMessage
              id='components.dataerror.click'
              defaultMessage='click here'
            />
          </Link>
          <Text size='18px' weight={300}>
            <FormattedMessage
              id='components.dataerror.refresh'
              defaultMessage=' to refresh.'
            />
          </Text>
        </TextGroup>
      )
    }
  }

  return (
    <Wrapper>
      <Empty>
        <Image name='empty-search' width='260px' />
        <Header size='18px' weight={500}>
          <FormattedMessage
            id='components.dataerror.header'
            defaultMessage='Oops, something went wrong here!'
          />
        </Header>
        {renderErrorHandling(message)}
      </Empty>
    </Wrapper>
  )
}

DataErrorContainer.propTypes = {
  onClick: PropTypes.func.isRequired
}

export default DataErrorContainer
