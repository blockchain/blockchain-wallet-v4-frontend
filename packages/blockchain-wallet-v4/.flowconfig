[ignore]

[include]

[libs]
./flow-typed
[lints]

[options]

[strict]
