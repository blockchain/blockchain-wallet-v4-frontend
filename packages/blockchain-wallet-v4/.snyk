# Snyk (https://snyk.io) policy file, patches or ignores known vulnerabilities.
version: v1.14.1
ignore: {}
# patches apply the minimum changes required to fix a vulnerability
patch:
  SNYK-JS-LODASH-567746:
    - stellar-sdk > lodash:
        patched: '2020-04-30T23:13:42.050Z'
    - stellar-sdk > stellar-base > lodash:
        patched: '2020-04-30T23:13:42.050Z'
    - stellar-sdk > stellar-base > js-xdr > lodash:
        patched: '2020-04-30T23:13:42.050Z'
