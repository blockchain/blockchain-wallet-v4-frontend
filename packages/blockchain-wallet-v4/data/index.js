import { Types } from '../src'

export { default as bchTx } from './tx.bch.json'
export { default as walletV1 } from './wallet.v1'
export { default as walletV2 } from './wallet.v2'
export { default as walletV3 } from './wallet.v3'
export { default as walletV3WithLegacy } from './wallet.v3-with-legacy'

export const createMockWalletState = wallet => ({
  walletPath: Types.Wrapper.fromJS({ wallet })
})

export const xpriv =
  'xprv9uMGUhxGeDoMcVjBXceA13ivqEUkfL7YzcFFnFWLj7San5jPi2cdPH144pNkyJuLMouf3acrvrtpQWJFGQLdReuWQwpMDvwLJcP7QEfysm8'
