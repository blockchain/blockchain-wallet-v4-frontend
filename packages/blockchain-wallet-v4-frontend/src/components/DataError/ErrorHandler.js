import React from 'react'
import { FormattedMessage } from 'react-intl'
import { prop } from 'ramda'
import styled from 'styled-components'

import { Button, Link, Text, TextGroup } from 'blockchain-info-components'
import { FETCH_FEES_FAILURE } from 'blockchain-wallet-v4/src/redux/payment/model'
import { checkForVulnerableAddressError } from 'services/misc'

import { BROKERAGE_INELIGIBLE, IneligibleErrorMessage } from '../../modals/components'
import { RECURRING_BUY_PERIOD_FETCH } from '../Flyout/model'

const MessageText = styled(Text)`
  width: 80%;
  margin-bottom: 20px;
`

const ErrorHandler = (props) => {
  const { message, onClick } = props
  const e2e = props['data-e2e']
  const errorMessage = prop('message', message) || prop('description', message)
  const vulnerableAddress = checkForVulnerableAddressError(message)
  if (vulnerableAddress) {
    return (
      <>
        <MessageText size='18px' weight={400}>
          {message}
        </MessageText>
        <Button nature='primary' onClick={() => onClick(vulnerableAddress)}>
          <Text size='18px' weight={400} color='white'>
            <FormattedMessage
              id='components.dataerror.archiveaddress'
              defaultMessage='Archive Address'
            />
          </Text>
        </Button>
      </>
    )
  }
  if (errorMessage === BROKERAGE_INELIGIBLE) {
    return <IneligibleErrorMessage />
  }
  if (errorMessage === FETCH_FEES_FAILURE) {
    return (
      <Text size='16px' weight={400}>
        <FormattedMessage
          id='components.dataerror.feesfetchfailure'
          defaultMessage='There was a problem fetching fees. Please try again later.'
        />
      </Text>
    )
  }
  if (errorMessage === RECURRING_BUY_PERIOD_FETCH) {
    return (
      <Text size='16px' weight={400} style={{ width: '300px' }}>
        <FormattedMessage
          id='modals.recurringbuys.period_fetch_error'
          defaultMessage='There was an error fetching recurring buy frequency options. Please try again.'
        />
      </Text>
    )
  }
  if (typeof errorMessage === 'string') {
    return (
      <Text size='16px' color='error' weight={500}>
        {errorMessage}
      </Text>
    )
  }
  return (
    <TextGroup inline>
      <Text size='18px' weight={400}>
        <FormattedMessage id='components.dataerror.body' defaultMessage='Please ' />
      </Text>
      <Link size='18px' data-e2e={e2e ? `${e2e}Link` : ''} onClick={onClick}>
        <FormattedMessage id='components.dataerror.click' defaultMessage='click here' />
      </Link>
      <Text size='18px' weight={400}>
        <FormattedMessage id='components.dataerror.refresh' defaultMessage=' to refresh.' />
      </Text>
    </TextGroup>
  )
}

export default ErrorHandler
