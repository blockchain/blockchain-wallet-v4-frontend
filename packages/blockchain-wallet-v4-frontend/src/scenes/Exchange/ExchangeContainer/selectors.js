import { selectors } from 'data'

export const getData = state => ({
  step: selectors.components.exchange.getStep(state)
})
