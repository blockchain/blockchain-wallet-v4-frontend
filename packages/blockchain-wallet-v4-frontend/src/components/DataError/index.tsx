import React from 'react'
import { FormattedMessage } from 'react-intl'
import { Image, Text } from 'blockchain-info-components'
import styled from 'styled-components'

import ErrorHandler from './ErrorHandler'

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
`
const Empty = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin-top: 20px;
`
const Header = styled(Text)`
  margin-top: 30px;
  margin-bottom: 20px;
`

const DataError = (props: {
  message?: { message: string | Error }
  onClick?: (a: any) => void
}) => {
  return (
    <Wrapper>
      <Empty>
        <Image name='empty-search' width='260px' />
        <Header size='18px' weight={500}>
          <FormattedMessage
            id='components.dataerror.header'
            defaultMessage='Oops, something went wrong here!'
          />
        </Header>
        <ErrorHandler {...props} />
      </Empty>
    </Wrapper>
  )
}

export default DataError
