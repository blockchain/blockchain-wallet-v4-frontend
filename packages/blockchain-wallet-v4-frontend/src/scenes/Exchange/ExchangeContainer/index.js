import React from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import { actions, model } from 'data'

import ExchangeForm from '../ExchangeForm'
import ExchangeConfirm from '../ExchangeConfirm'
import { getData } from './selectors'

const { EXCHANGE_STEPS } = model.components.exchange

export class ExchangeContainer extends React.PureComponent {
  componentDidMount () {
    this.props.profileActions.fetchUser()
    this.props.actions.setStep(EXCHANGE_STEPS.EXCHANGE_FORM)
  }

  componentDidUpdate () {
    // Clear location state when going to confirm screen
    // So that it won't force swap when going back to exchange
    if (this.props.step === EXCHANGE_STEPS.CONFIRM)
      this.props.routerActions.replace('/swap', {})
  }

  componentWillUnmount () {
    this.props.actions.clearSubscriptions()
  }

  render () {
    const { step, from, to, fix, amount } = this.props
    switch (step) {
      case EXCHANGE_STEPS.EXCHANGE_FORM:
        return <ExchangeForm {...{ from, to, fix, amount }} />
      case EXCHANGE_STEPS.CONFIRM:
        return <ExchangeConfirm />
      default:
        return <ExchangeForm />
    }
  }
}

ExchangeContainer.propTypes = {
  step: PropTypes.number.isRequired
}

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions.components.exchange, dispatch),
  routerActions: bindActionCreators(actions.router, dispatch),
  profileActions: bindActionCreators(actions.modules.profile, dispatch)
})

export default connect(
  getData,
  mapDispatchToProps
)(ExchangeContainer)
