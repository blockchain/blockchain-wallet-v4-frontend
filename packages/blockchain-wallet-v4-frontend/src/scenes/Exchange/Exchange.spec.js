import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { Remote } from 'blockchain-wallet-v4'
import { ExchangeScene } from './index'
import ExchangeContainer from './ExchangeContainer'
import DataError from 'components/DataError'

jest.mock('blockchain-info-components', () => ({
  BlockchainLoader: 'BlockchainLoader',
  Button: 'Button',
  FlatLoader: 'FlatLoader',
  Icon: 'Icon',
  Image: 'Image',
  Link: 'Link',
  Text: 'Text',
  TextInput: 'TextInput',
  TextGroup: 'TextGroup'
}))

const FROM = 'BTC'
const TO = 'ETH'

const props = {
  hasEmail: true,
  logEnterExchange: jest.fn(),
  fetchUser: jest.fn(),
  location: {
    state: {
      from: FROM,
      to: TO
    }
  }
}

describe('ExchangeScene', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  it('renders email required screen correctly', () => {
    const component = shallow(<ExchangeScene {...props} hasEmail={false} />)
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
  it('calls fetchUser at start', () => {
    const component = shallow(<ExchangeScene {...props} hasEmail={false} />)
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
  it('renders loading correctly', () => {
    const component = shallow(
      <ExchangeScene userCreated={Remote.Loading} {...props} />
    )
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
  it('renders not asked correctly', () => {
    const component = shallow(
      <ExchangeScene userCreated={Remote.NotAsked} {...props} />
    )
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
  it('renders exchange correctly', () => {
    const component = shallow(
      <ExchangeScene userCreated={Remote.Success(true)} {...props} />
    )
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
    const exchange = component.find(ExchangeContainer)
    expect(exchange.prop('from')).toBe(FROM)
    expect(exchange.prop('to')).toBe(TO)
  })
  it('renders getstarted correctly', () => {
    const component = shallow(
      <ExchangeScene userCreated={Remote.Success(false)} {...props} />
    )
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
  it('renders failure correctly', () => {
    const component = shallow(
      <ExchangeScene userCreated={Remote.Failure({})} {...props} />
    )
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
    const dataError = component.find(DataError)
    dataError.prop('onClick')()
    expect(props.fetchUser).toHaveBeenCalledTimes(1)
  })
})
