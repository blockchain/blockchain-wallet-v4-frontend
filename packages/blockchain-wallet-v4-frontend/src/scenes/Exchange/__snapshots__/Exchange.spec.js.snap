// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeScene calls fetchUser at start 1`] = `<EmailRequired />`;

exports[`ExchangeScene renders email required screen correctly 1`] = `<EmailRequired />`;

exports[`ExchangeScene renders exchange correctly 1`] = `
<Exchange__Wrapper>
  <Exchange__Container>
    <Exchange__Column>
      <Connect(ExchangeContainer)
        from="BTC"
        to="ETH"
      />
    </Exchange__Column>
  </Exchange__Container>
</Exchange__Wrapper>
`;

exports[`ExchangeScene renders failure correctly 1`] = `
<DataErrorContainer
  onClick={[MockFunction]}
/>
`;

exports[`ExchangeScene renders getstarted correctly 1`] = `
<Exchange__Wrapper>
  <GetStarted />
</Exchange__Wrapper>
`;

exports[`ExchangeScene renders loading correctly 1`] = `
<Exchange__Wrapper>
  <BlockchainLoader
    height="200px"
    width="200px"
  />
</Exchange__Wrapper>
`;

exports[`ExchangeScene renders not asked correctly 1`] = `
<Exchange__Wrapper>
  <BlockchainLoader
    height="200px"
    width="200px"
  />
</Exchange__Wrapper>
`;
